libbeat/processors/actions/cloud_metadata.go

```diff
@@ -0,0 +1,224 @@
+		out := common.MapStr{}
+
+		if instance, ok := m["instance"].(map[string]interface{}); ok {
+			s.Schema{
+				"instance_id":       c.StrFromNum("id"),
+				"instance_type":     c.Str("machineType"),
+				"availability_zone": c.Str("zone"),
+			}.ApplyTo(out, instance)
+		}
+
+		if project, ok := m["project"].(map[string]interface{}); ok {
+			s.Schema{
+				"project_id": c.Str("projectId"),
+			}.ApplyTo(out, project)
+		}
+
+		return out
+	}
+)
+
+// init registers the cloud_metadata processor.
+func init() {
+	processors.RegisterPlugin("cloud_metadata",
+		configChecked(newCloudMetadata,
+			requireFields(),
+			allowedFields("when")))
+}
+
+// result is the result of a query for a specific hosting service's metadata.
+type result struct {
+	provider string        // Hosting provider type.
+	err      error         // Error that occured while fetching (if any).
+	metadata common.MapStr // A specific subset of the metadata received from the hosting provider.
+}
+
+// fetchJSON query metadata from a hosting provider's metadata service.
+func fetchJSON(
+	provider string,
+	headers map[string]string,
+	url string,
+	conv func(map[string]interface{}) common.MapStr,
+) result {
+	result := result{provider: provider}
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		result.err = errors.Wrapf(err, "failed to create http request for %v", provider)
+		return result
+	}
+	for k, v := range headers {
+		req.Header.Add(k, v)
+	}
+
+	rsp, err := metaClient.Do(req)
+	if err != nil {
+		result.err = errors.Wrapf(err, "failed requesting %v metadata", provider)
+		return result
+	}
+	defer rsp.Body.Close()
+
+	if rsp.StatusCode != http.StatusOK {
+		result.err = errors.Errorf("failed with http status code %v", rsp.StatusCode)
+		return result
+	}
+
+	all, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		result.err = errors.Wrapf(err, "failed requesting %v metadata", provider)
+		return result
+	}
+
+	// Decode JSON.
+	dec := json.NewDecoder(bytes.NewReader(all))
+	dec.UseNumber()
+	err = dec.Decode(&result.metadata)
+	if err != nil {
+		result.err = errors.Wrapf(err, "failed to unmarshal %v JSON of '%v'", provider, string(all))
+		return result
+	}
+
+	data := conv(result.metadata)
+	data["provider"] = provider
+	result.metadata = common.MapStr{"cloud": data}
+
+	return result
+}
+
+// fetchMetadata attempts to fetch metadata in parallel from each of the
+// hosting providers supported by this processor. It wait for the results to
+// be returned or for a timeout to occur then returns the results that
+// completed in time.
+func fetchMetadata() []result {
+	c := make(chan result)
+	go func() { c <- fetchJSON("digitalocean", nil, digitaloceanMetadataURL, digitaloceanSchema) }()
+	go func() { c <- fetchJSON("ec2", nil, ec2MetadataURL, ec2Schema) }()
+	go func() { c <- fetchJSON("gce", gceHeaders, gceMetadataURL, gceSchema) }()
+
+	var results []result
+	timeout := time.After(5 * time.Second)
```

Comment:
The timeout is now configurable and defaults to 3s. In practice when running on EC2 the request is completed in 2ms.